sstables/index_reader: abort reading during shutdown

lkshminarayanan opened the pull request:
This PR adds support for aborting index reads from within `index_consume_entry_context::consume_input` when the server is being stopped. The abort source is now propagated down to the `index_consume_entry_context`, making it available for `consume_input` to check if an abort has been requested. If an abort is detected, `consume_input` will throw an exception to stop the index read operation.

lkshminarayanan commented:
PR is based on this comment - https://github.com/scylladb/scylladb/pull/19190#discussion_r1635176907. 

Backport to 5.4 and 6.0, so that #19190 get the benefit.

denesb reviewed (changes requested):
My main comment is to replace `std::optional<std::reference_wrapper<abort_source>>` with `abort_source*`. It means the same thing and is much easier on the eye.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9924/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9924/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2243/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10264/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9924/)
#### Build Details:
- Duration: 6 hr 3 min
- Builder: i-0ab97e6f64a8cc126 (m5ad.8xlarge)

lkshminarayanan commented on replica/database.cc:
```diff
@@ -313,7 +313,8 @@ class db_user_types_storage : public data_dictionary::dummy_user_types_storage {
 database::database(const db::config& cfg, database_config dbcfg, service::migration_notifier& mn, gms::feature_service& feat, const locator::shared_token_metadata& stm,
-        compaction_manager& cm, sstables::storage_manager& sstm, lang::manager& langm, sstables::directory_semaphore& sst_dir_sem, utils::cross_shard_barrier barrier)
+        compaction_manager& cm, sstables::storage_manager& sstm, lang::manager& langm, sstables::directory_semaphore& sst_dir_sem, utils::cross_shard_barrier barrier,
+        std::optional<std::reference_wrapper<abort_source>> abort)
```
I have updated the PR to use `abort_source*`.

lkshminarayanan commented:
V2 Changelog :
 - Use `abort_source*` instead of `std::optional<std::reference_wrapper<abort_source>>`

avikivity commented:
I don't think we should backport this or #19190. It doesn't fix a regression. Improvements will reach users via the next release.

avikivity commented on sstables/index_reader.hh:
```diff
@@ -318,11 +319,12 @@ public:
     index_consume_entry_context(const sstable& sst, reader_permit permit, IndexConsumer& consumer, trust_promoted_index trust_pi,
             input_stream<char>&& input, uint64_t start, uint64_t maxlen,
-            std::optional<column_values_fixed_lengths> ck_values_fixed_lengths, tracing::trace_state_ptr trace_state = {})
+            std::optional<column_values_fixed_lengths> ck_values_fixed_lengths, tracing::trace_state_ptr trace_state = {}, abort_source* abort = nullptr)
```
What callers don't supply it?

It those are just tests, maybe it's better to make it mandatory and have the tests supply a dummy abort_source. Default parameters are an invitation to missing parameters.

lkshminarayanan replied:
Now the index_consume_entry_context gets the abort_source directly from the sstable/sstable_manager. Updated the tests as well.

avikivity commented on sstables/index_reader.hh:
```diff
@@ -186,6 +186,11 @@ public:
+            sstlog.warn("Aborting index consumer. Reason : %s", _abort->abort_requested_exception_ptr());
+            return proceed::no;
+        }
+
```
We can't just stop like that. We have to return something, no? How will the caller proceed? It thinks it needs to wait for I/O, but no I/O is forthcoming.

raphaelsc replied:
I think we may have to throw here. It seems proceed::no can cause input_stream<CharType>::consume() to end prematurely (silently), which apparently may result in sstable writer generating a bad filter in shutdown (because it (wrongly) thinks the index was entirely consumed). With the exception, we're aborting the process which initiated the consumption of index (e.g. regeneration of summary or filter).

raphaelsc replied:
It's not the end of the world if we lose a sstable write in shutdown. That happens today with compactions that are almost completed but do get aborted.

lkshminarayanan replied:
@raphaelsc That's what I ended up doing. Instead of returning `proceed:no`, the exception_ptr in abort source is now rethrown. I was actually concerned about allowing the exception to propagate up causing the write to fail, so I was trying to handle it in `maybe_rebuild_filter_from_index()`. But what you say makes sense, so I have removed the exception handling and the write fails when the index reader gets aborted.

One thing to note is, `maybe_rebuild_filter_from_index()` is also in the path of memtable flush to disk but although the filter will never be rebuilt during that flow as the filter will always be optimal. So hopefully this is okay.

raphaelsc replied:
from exception handling perspective, we don't retry flush on shutdown, so we'll have to replay later through commitlog. So we're good here.

As for not stalling flush with the rebuild from index (because that can result in memory pressure. that's what you're concerned right?), maybe it can be a problem with twcs, which segregates memtable data into multiple write streams, one for each window involved. let's say that memtable data spans 2 windows, where 99% of it belongs to a single window. Estimation average the input partition count by number of windows spanned, so we have suboptimal filters for both. Maybe we should bypass maybe_rebuild_filter_from_index() if sstable origin == "memtable", meaning we're favoring flush time (decreasing mem pressure as we're only able to clear memtable after sealing sstable), and sstables produced by flush have a size proportional to memory size, so overestimated filters for them won't play a large role in memory usage, and they can later be adjusted through compactions. what do you think?

lkshminarayanan replied:
> from exception handling perspective, we don't retry flush on shutdown, so we'll have to replay later through commitlog. So we're good here.

Nice! This is what I was actually concerned about. Although the chances of a filter rebuild happening during a memtable flush are almost zero, I was concerned about missing any edge cases that I'm not aware of which could produce sub-optimal filters triggering a filter rebuild. It's good to know that commitlog replay will take care of failures during such cases.

lkshminarayanan replied:
> As for not stalling flush with the rebuild from index (because that can result in memory pressure. that's what you're concerned right?), maybe it can be a problem with twcs, which segregates memtable data into multiple write streams, one for each window involved. let's say that memtable data spans 2 windows, where 99% of it belongs to a single window. Estimation average the input partition count by number of windows spanned, so we have suboptimal filters for both. Maybe we should bypass maybe_rebuild_filter_from_index() if sstable origin == "memtable", meaning we're favoring flush time (decreasing mem pressure as we're only able to clear memtable after sealing sstable), and sstables produced by flush have a size proportional to memory size, so overestimated filters for them won't play a large role in memory usage, and they can later be adjusted through compactions. what do you think?

This should be totally fine! One question though : with other strategies like LCS and STCS, during memtable flush, the exact partition count is known and is passed to bloom filter when it is created. So they always turn out to be optimal when checked inside `maybe_rebuild_filter_from_index()` and the filter is never rebuilt. IIUC, with TWCS, individual memtables within each windows will be flushed to their own separate sstables. In that case, why do we have to use a partition estimate to create bloom filters? Can't we use the same logic as LCS and STCS to create filters with the exact partition count?

raphaelsc replied:
> > As for not stalling flush with the rebuild from index (because that can result in memory pressure. that's what you're concerned right?), maybe it can be a problem with twcs, which segregates memtable data into multiple write streams, one for each window involved. let's say that memtable data spans 2 windows, where 99% of it belongs to a single window. Estimation average the input partition count by number of windows spanned, so we have suboptimal filters for both. Maybe we should bypass maybe_rebuild_filter_from_index() if sstable origin == "memtable", meaning we're favoring flush time (decreasing mem pressure as we're only able to clear memtable after sealing sstable), and sstables produced by flush have a size proportional to memory size, so overestimated filters for them won't play a large role in memory usage, and they can later be adjusted through compactions. what do you think?
> 
> This should be totally fine! One question though : with other strategies like LCS and STCS, during memtable flush, the exact partition count is known and is passed to bloom filter when it is created. So they always turn out to be optimal when checked inside `maybe_rebuild_filter_from_index()` and the filter is never rebuilt. IIUC, with TWCS, individual memtables within each windows will be flushed to their own separate sstables. In that case, why do we have to use a partition estimate to create bloom filters? Can't we use the same logic as LCS and STCS to create filters with the exact partition count?

the problem exists because we have a single memtable, which will be flushed into multiple sstables. It's 1:N, not 1:1. for LCS and ICS, we don't have such a segregator on flush, just during compaction.

raphaelsc replied:
> > > As for not stalling flush with the rebuild from index (because that can result in memory pressure. that's what you're concerned right?), maybe it can be a problem with twcs, which segregates memtable data into multiple write streams, one for each window involved. let's say that memtable data spans 2 windows, where 99% of it belongs to a single window. Estimation average the input partition count by number of windows spanned, so we have suboptimal filters for both. Maybe we should bypass maybe_rebuild_filter_from_index() if sstable origin == "memtable", meaning we're favoring flush time (decreasing mem pressure as we're only able to clear memtable after sealing sstable), and sstables produced by flush have a size proportional to memory size, so overestimated filters for them won't play a large role in memory usage, and they can later be adjusted through compactions. what do you think?
> > 
> > This should be totally fine! One question though : with other strategies like LCS and STCS, during memtable flush, the exact partition count is known and is passed to bloom filter when it is created. So they always turn out to be optimal when checked inside `maybe_rebuild_filter_from_index()` and the filter is never rebuilt. IIUC, with TWCS, individual memtables within each windows will be flushed to their own separate sstables. In that case, why do we have to use a partition estimate to create bloom filters? Can't we use the same logic as LCS and STCS to create filters with the exact partition count?
> 
> the problem exists because we have a single memtable, which will be flushed into multiple sstables. It's 1:N, not 1:1. for LCS and ICS, we don't have such a segregator on flush, just during compaction.

Maybe it's possible to feed an exact estimation with twcs by having a partition count tracker which impl is compaction strategy specific (that lives in memtable) and we feed it into the splitter since we want each stream to know exactly the amount of partitions it will have to consume for creation of the filter. But it may be some significant amount of work and I don't see we doing it any time soon.

lkshminarayanan replied:
Okay.

lkshminarayanan replied:
Allowing the write fail, propagates the exception and the server fails due to unhandled exception. So, I have updated the PR to drop the old filter only after the new one is built during rebuild. If the rebuild is aborted, the old filter still remains and will be written to disk.

avikivity commented:
test?

Won't be easy, perhaps you'll need to add wait injection to some file wrapper.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9947/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9947/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10280/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2264/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9947/)
#### Build Details:
- Duration: 6 hr 8 min
- Builder: i-035d70b0099a94498 (m5ad.8xlarge)

lkshminarayanan commented:
V3 Changelog :
 - `abort_source` is not an optional parameter anymore and all methods pass down abort_source to sstable_manager. 
 - added a new method `sstables_manager::get_abort_source` to return abort_source reference from the manager
 - `index_consume_entry_context` instead of getting `abort_source` via constructor parameter, now directly gets it from the sstable_manager.
 - added a test case that aborts and expects the `index_consume_entry_context::consume_input` to throw an exception.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10056/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10056/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/index_reader_test
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10056/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10056/)
:x: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2337/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10361/)
#### Failed Tests (114/31010):
- [test_auth_v2_during_recovery.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10056/testReport/junit/%28root%29/test_auth_v2_migration/Tests___Unit_Tests___test_auth_v2_during_recovery_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_auth_v2_during_recovery.debug.3)
- [test_service_levels_work_during_recovery.dev.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10056/testReport/junit/%28root%29/test_raft_service_levels/Tests___Unit_Tests___test_service_levels_work_during_recovery_dev_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_service_levels_work_during_recovery.dev.3)
- [test_mutation_schema_change.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10056/testReport/junit/%28root%29/test_mutation_schema_change/Tests___Unit_Tests___test_mutation_schema_change_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_mutation_schema_change.debug.1)
- [test_reboot.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10056/testReport/junit/%28root%29/test_commitlog/Tests___Unit_Tests___test_reboot_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_reboot.debug.2)
- [test_feature_deprecation_works.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10056/testReport/junit/%28root%29/test_deprecating_cluster_features/Tests___Unit_Tests___test_feature_deprecation_works_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_feature_deprecation_works.debug.2)
- [test_feature_deprecation_works.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10056/testReport/junit/%28root%29/test_deprecating_cluster_features/Tests___Unit_Tests___test_feature_deprecation_works_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_feature_deprecation_works.debug.3)
- [test_raft_recovery_basic.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10056/testReport/junit/%28root%29/test_raft_recovery_basic/Tests___Unit_Tests___test_raft_recovery_basic_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_raft_recovery_basic.debug.1)
- [test_raft_recovery_basic.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10056/testReport/junit/%28root%29/test_raft_recovery_basic/Tests___Unit_Tests___test_raft_recovery_basic_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_raft_recovery_basic.debug.3)
- [test_recover_stuck_raft_recovery.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10056/testReport/junit/%28root%29/test_raft_recovery_stuck/Tests___Unit_Tests___test_recover_stuck_raft_recovery_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_recover_stuck_raft_recovery.debug.3)
- [test_quorum_lost_during_node_join_response_handler.dev.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10056/testReport/junit/%28root%29/test_raft_no_quorum/Tests___Unit_Tests___test_quorum_lost_during_node_join_response_handler_dev_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_quorum_lost_during_node_join_response_handler.dev.1)
#### Build Details:
- Duration: 10 hr
- Builder: i-07906d91822024100 (m5ad.8xlarge)

lkshminarayanan commented:
V4 Changelog :
 - Drop the old filter in `sstable::maybe_rebuild_filter_from_index` only after the new filter is built. This way, if the rebuild gets aborted, we still have the old filter (although not optimal) to write to disk.

lkshminarayanan commented:
V5 Changelog :
 - Do not abort index reading when `index_consume_entry_context` is called from the `index_reader`.
 
 More info on this :
 
` index_consume_entry_context` has exactly three callers:
 - `sstable::generate_summary`
 - `sstable::maybe_rebuild_filter_from_index`
 - `sstable::index_reader`
 
 The first two methods can handle any exception thrown when the index is consumed by `consumer_ctx.consume_input()`. So, this PR now passes proper abort source to `index_consume_entry_context` from these two methods. Note that with `sstable::maybe_rebuild_filter_from_index` we do end up with a suboptimal filter if it gets aborted midway.
 
 However, the `sstable::index_reader` class has a hierarchy of callers and to update them all to handle the exception thrown during index read is not trivial. Not handling these exceptions causes the server to crash during shutdown as the exception propagates up and is considered an unhandled exception. This is why all the various tests were failing previously in this PR. 

So, this PR sends a dummy abort source from the `sstable::index_reader` when creating an `index_consume_entry_context`.  If required, we could pass the proper abort source from the `index_reader` and update its callers to handle exception in a case by case basis in future PRs.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10181/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10181/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/index_reader_test
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10181/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2431/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10456/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10181/)
#### Build Details:
- Duration: 2 hr 38 min
- Builder: spider1.cloudius-systems.com

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10221/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10221/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/index_reader_test
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10221/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10489/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2464/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10221/)
#### Build Details:
- Duration: 2 hr 39 min
- Builder: spider5.cloudius-systems.com

avikivity commented on sstables/index_reader.hh:
```diff
@@ -186,6 +186,12 @@ public:
+            auto ex = _abort.abort_requested_exception_ptr();
+            sstlog.warn("aborting index_consume_entry_context::process_state() : {}", ex);
+            std::rethrow_exception(ex);
+        }
```
Why the warning here? What's the user going to do about it?

I understand you wanted to see it during debugging, but now, it's just noise. You can all abort_source::check().

lkshminarayanan replied:
I agree, I have removed the warning and updated the code.

lkshminarayanan commented:
V6 Changelog :
 - use abort_source.check() and removed a warning that was emitted when exception was thrown.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10379/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10379/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/index_reader_test
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10379/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10605/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2586/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10379/)
#### Build Details:
- Duration: 2 hr 38 min
- Builder: spider2.cloudius-systems.com

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10413/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10413/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/index_reader_test
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10413/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2615/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10634/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10413/)
#### Build Details:
- Duration: 6 hr 3 min
- Builder: i-0c57e916456bc12fb (m5ad.8xlarge)
